Pilight receive match fix for bug 4637

janLo opened the pull request:
**Description:**
This is a bugfix for Bug #4637 

It changes the matching of received codes to fist check that the protocol is within the configured protocols and then match if the rest of the configured values are found in the received data.

**Related issue (if applicable):** fixes #4637 

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51

DavidLP commented:
This is not a good solution. We should just not check the received code and not force type casts with voluptuous. Let me recap how this bug was introduced:

- The C library of pilight has a bug that when a _protocol_ info is not in a list is crashes with a segfault ([link](https://github.com/pilight/pilight/issues/296)). Thus after feedback during the first comit of the pilight component (dc9f990) a voluptuous scheme was put in place to ensure that the send _protocol_ is in a list. This scheme is called `RF_CODE_SCHEMA` ([link](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/pilight.py#L40)). In retrospect the name is not the best, because it is for sending data, not receiveing. The name `RF_SND_CODE_SCHEMA` would be better.
- In comit 4484a7a at [this line](https://github.com/home-assistant/home-assistant/commit/4484a7a94b8ba91f77b60df2bb35a944cb164e7f#diff-46a0d82f81fd2a1e0ed19abf33c14a56R27) the received code scheme was set to the send code scheme, wich is not correct. Because while the sending protocol info has to be in a list, the received protocol is not in a list.
- Thus this [check](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/switch/pilight.py#L115) is never true introducing the bug #4637

The clean solution is not to check the received code with voluptuous instead of adding a [special protocol handling](https://github.com/home-assistant/home-assistant/pull/4639/files#diff-46a0d82f81fd2a1e0ed19abf33c14a56R73) case. 

A quick fix is to take the original code and just change [this](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/switch/pilight.py#L27) to:

```
COMMAND_SCHEMA = vol.Schema({
    vol.Optional(CONF_PROTOCOL): cv.string
}, extra=vol.ALLOW_EXTRA)

```

And everything is working again.

DavidLP commented:
While I added full coverage unit testing for the pilight hub, the pilight switch is lacking any tests. To prevent these kind of regressions in the future a unit test for the switch would be great.

janLo commented:
Ok, I thought it was a feature to match on more than one protocol. I'll adjust the PR.

janLo commented:
Simply changing the type back will reintroduce bug #3355. So we can either copy the data from the mapping proxy to a regular dict on every service call or map the data for received codes as done in the commits above. Both approaches are not perfect but I think the latter one is less intrusive because it only affects the receiving in the switch component (that will not work at all otherwise).

Of course, we can also convert the protocol in the received data to a list before we match. But this will break all event triggers that other users might have using a protocol as match data.

I would suggest to take this PR as hotfix until we have unittests in place to find a solution that definitely work in all cases.

janLo commented:
@DavidLP can you agree with that?

DavidLP commented:
I have the changes I mentioned above running at home. I do not see #3355. Do you? 
The mapping proxy issue arises only, when we change data, but we do not do that.

Maybe we miscommunicated. I mean _just_ my change from above, nothing else from the PR. Then everything is fine. So it is really a small bugfix.

janLo commented:
@DavidLP  don't we have to change the data before sending it to the pilight daemon to prevent the crash that you described in https://github.com/pilight/pilight/issues/296 - and that caused the line that you introduced that caused #3355 :  https://github.com/home-assistant/home-assistant/commit/dd4611064f907b567f4865219ddc8c0a96ddfd7d#diff-6e9c4d5f1e7b9e71660f62bd3728f53fL79 ?

If I have a config like this:
```YAML
My Switch:
  on_code:
    protocol: brennenstuhl
    unitcode: 1
    systemcode: 31
    'on': 1
  on_code_receive:
    protocol: elro_800_switch
    unitcode: 1
    systemcode: 1
    state: 'on'
```

Then someone has to wrap the protocol into a list before sending to not cause https://github.com/pilight/pilight/issues/296. There are two possibilities:
* If this is done in the voloptuous code, then we have to convert it back before matching received codes OR use this PR as the codes that come from pilight does not have the protocol wrapped in a list. 
* If this is done in the send_code method, then we have to deal with the mappingprody issue described in #3351 and fixed in #3355

What do I miss here? Is there another place where https://github.com/pilight/pilight/issues/296 is already addressed?

DavidLP commented:
@janLo: Let me try to explain:

1. There is a [pilight hub component](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/pilight.py) that is working fine and has full coverage unit testing. The [pilight hub component](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/pilight.py) uses a [python package](https://github.com/DavidLP/pilight) that communicates via a socket connection to the [pilight daemon](https://wiki.pilight.org/doku.php/pdaemon) that in turn uses the [pilight c-library](https://wiki.pilight.org/doku.php/start). So the communication in home assistant is always done via the [pilight hub component](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/pilight.py). In this component the data patching is implemented to ensure that the _protocol_ info is in a list. This component should not be changed.

2. The communication protocol from the [pilight hub component](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/pilight.py) via the [python package](https://github.com/DavidLP/pilight) to the [pilight daemon](https://wiki.pilight.org/doku.php/pdaemon) are JSON objects. So everything that is send has to be [serializable](https://github.com/DavidLP/pilight/blob/master/pilight/pilight.py#L177) by the python JSON library. In turn everything that is received are JSON objects. This data is put on the [home assistant event bus](https://home-assistant.io/developers/architecture/) and can be used by other components (switch, automation). Since a lot of different data types can be serialized by JSON and every device defines its own protocol, we cannot and should not define any protocol specifics in the [pilight hub component](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/pilight.py) and anywhere else.

3. There is a [pilight switch component](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/switch/pilight.py#L25) that uses the [pilight hub component](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/pilight.py) to send / receive data needed to implement a switch. The switch component talks to the hub via the home assistant event bus and is never talking directly to the [pilight daemon](https://wiki.pilight.org/doku.php/pdaemon). So do not worry about any old bug since the proper interface is ensured by the [pilight hub component](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/pilight.py).

So as a result just change the 3 lines in the pilight switch as I proposed above and you will see it works.

janLo commented:
@DavidLP Thanks for explaining. But the point I missed was, that the data-patching that was gone in https://github.com/home-assistant/home-assistant/commit/dd4611064f907b567f4865219ddc8c0a96ddfd7d#diff-6e9c4d5f1e7b9e71660f62bd3728f53fL79 and I didn't see who does it now. Now I realized that it was  replaced by this shema specification: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/pilight.py#L98.

janLo committed:
Pilight: dont protocol as list in COMMAND_SCHEMA

As described in bug #4637 the protocol should not be wrapped in a list
in the spec of COMMAND_SCHEMA because this causes the component to
never successfully match any received rf code.

As pointed ot in PR #4639 the easiest way to do this, is to not derive
COMMAND_SCHEMA from RF_CODE_SCHEMA and specify protocol as simple
string there.

This fixes bug #4637.

Signed-off-by: Jan Losinski <losinski@wh2.tu-dresden.de>

janLo committed:
Pilight: Add "unitcode" to command schema.

This adds "unitcode" to the COMMAND_SCHEMA. It is used for example in
the brennenstuhl protocol of pilight.

Signed-off-by: Jan Losinski <losinski@wh2.tu-dresden.de>

balloob commented:
@DavidLP, let us know when this is ok to merge 👍

DavidLP commented on homeassistant/components/switch/pilight.py:
```diff
@@ -21,13 +22,16 @@
     vol.Optional('on'): cv.positive_int,
     vol.Optional('off'): cv.positive_int,
     vol.Optional(CONF_UNIT): cv.positive_int,
+    vol.Optional(CONF_UNITCODE): cv.positive_int,
```
Sorry for the late reply, I did not know that a review can be just pending without submitting. I had it pending for days...

DavidLP replied:
Unfortunately the _most likely_ turned out to be false.

janLo replied:
"unitcode" is still used only as an int.

CurlyMoo commented:
If you think we can improve pilight in whatever way to make it easier to interface with, please don't hesitate to join us on our forum to discuss it.

janLo commented:
@CurlyMoo merging of PR https://github.com/pilight/pilight/pull/322 would be a nice start, so that we can build protocol validators from this data. Thanks.
